### ISSUE TITLE
FileNotFoundError: [Errno 2] No such file or directory - ffmpeg audio/video muxing error

### ISSUE THREAD

**Issue by ymgenesis** [aliparlakci/bulk-downloader-for-reddit#106]
Starting to test out the new 1.8.0 release on macOS and Linux.

**System**
macOS 10.14.6
ffmpeg version:
```
ffmpeg version 4.2.3 Copyright (c) 2000-2020 the FFmpeg developers
  built with Apple clang version 11.0.0 (clang-1100.0.33.17)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/4.2.3_1 --enable-shared --enable-pthreads --enable-version3 --enable-avresample --cc=clang --host-cflags=-fno-stack-check --host-ldflags= --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libbluray --enable-libdav1d --enable-libmp3lame --enable-libopus --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-librtmp --enable-libspeex --enable-libsoxr --enable-videotoolbox --disable-libjack --disable-indev=jack
  libavutil      56. 31.100 / 56. 31.100
  libavcodec     58. 54.100 / 58. 54.100
  libavformat    58. 29.100 / 58. 29.100
  libavdevice    58.  8.100 / 58.  8.100
  libavfilter     7. 57.100 /  7. 57.100
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  5.100 /  5.  5.100
  libswresample   3.  5.100 /  3.  5.100
  libpostproc    55.  5.100 / 55.  5.100
```

**Issue**

When downloading v.redd.it videos, temporary `_audio` and `_video` files download, but fail to mux. The following exception is thrown:

`FileNotFoundError: [Errno 2] No such file or directory`

In the download directory I'm left with a bunch of `_audio` and `_video` files.

Reading the logs, I suspect it has something to do with the way the output file is named in the ffmpeg command. I tried doing the muxing of two `_audio` and `_video` files with the same ffmpeg command, with just an output name of 'test.mp4', and it was successful.

**Logs**
[CONSOLE_LOG.txt, FAILED.json, and POSTS.json](https://pastebin.com/cQ41sjnp)

PS: The new --no-dupes & --downloaded-posts options are amazing. Thank you!

**Comment by aliparlakci** [comment]
I have just figured that it was because of the strange characters in the title of the posts. The posts titles have bothered me beyond recognition.

I will try to fix it ASAP. You can exclude the title in the filenames for quick fix.

**Comment by ymgenesis** [comment]
> I have just figured that it was because of the strange characters in the title of the posts. The posts titles have bothered me beyond recognition.
> 
> I will try to fix it ASAP. You can exclude the title in the filenames for quick fix.

I hear ya. Thanks for the tip, and your work!

**Comment by aliparlakci** [comment]
Can you use the latest commit on the develop branch and see if it helps?

**Comment by ymgenesis** [comment]
> Can you use the latest commit on the develop branch and see if it helps?

I'll give it a shot and get back to you!

**Comment by aliparlakci** [comment]
I did not see that git failed to push to the develop branch :D Sorry for that.

It, hopefully, should be live on the develop branch now.

**Comment by ymgenesis** [comment]
> I did not see that git failed to push to the develop branch :D Sorry for that.
> 
> It, hopefully, should be live on the develop branch now.

Unfortunately I get the same type of errors.
[Pastebin](https://pastebin.com/aFW90jmB)

**Comment by aliparlakci** [comment]
I cannot figure out problem right now. Until I solve the problem, you can add the following line in the src/vreddit.py after the 33th line:
```python
filename = post['POSTID']+".mp4"
```

It should look like this:

```python
getFile(videoName,videoName,directory,videoURL,silent=True)
getFile(audioName,audioName,directory,audioURL,silent=True)
filename = post['POSTID']+".mp4"
try:
    self._mergeAudio(videoName,
                     audioName,
                     filename,
```

**Comment by aliparlakci** [comment]
> And simply named a random file that, and macOS accepted the filename without giving an error. Maybe it has something to do with how the &'s or periods are being processed in python, or the way it's executed in the code.

It says in this site that MacOS has no limitations on characters: https://blackbeltreview.wordpress.com/2015/01/27/illegal-filename-characters-on-windows-vs-mac-os/

**Comment by ymgenesis** [comment]
Right! It used to, I guess I never figured out that now you can put / and :
Anyways, no worries for now! Thanks for looking into it.

**Comment by aliparlakci** [comment]
Can you create this file `/path/to/dl/PublicFreakout/SUPERGUESSOUS_U.S._security_forces_hunt_down_journalists_covering_GeorgeFloyd_protests._VICE_reporter_@MichaelAdams317_plea“I’m_Press!_Press!_Press!”_as_he's_thrown_to_the_ground,_beaten,_and_pepper-sprayed_directl_gu5yru.mp4` and see if it raises any errors?

If it does, can you remove the strange characters one by one and see when it accepts the filename?

**Comment by ymgenesis** [comment]
I used `touch` (macOS) to create a file with the following filename:
`SUPERGUESSOUS_U.S._security_forces_hunt_down_journalists_covering_GeorgeFloyd_protests._VICE_reporter_@MichaelAdams317_plea“I’m_Press!_Press!_Press!”_as_he's_thrown_to_the_ground,_beaten,_and_pepper-sprayed_directl_gu5yru.mp4`

the file was crated with the following name, cutting off before the quotes:
`SUPERGUESSOUS_U.S._security_forces_hunt_down_journalists_covering_GeorgeFloyd_protests._VICE_reporter_@MichaelAdams317_plea“`

I had to remove the `“` (sideways double quotes), `’` (sideways single quote), and `!` (regular exclamation mark), and the file creation worked. I would suggest removing `@` (regular at) as well, as it means many things in Linux and Unix (mac) OSs. Maybe even the periods `.`, other than the one before the extension.

Notice the difference between `“` (slanted) and `"` (regular), and `’` (slanted) `'` regular (not a backtick `\``. I can't seem to find information on the differences, but I know they're each individual characters.

**Comment by ymgenesis** [comment]
Did some more probing. Here's an error:
`FileNotFoundError: [Errno 2] No such file or directory: 'ffmpeg -i /path/to/gvuqt5_audio -i /path/to/gvuqt5_video -c:v copy -c:a aac -strict experimental /path/to/ReformTheUS_Downtown_NYC_-_8_20pm_-_Police_advancing_and_beating_peaceful_bicyclist_protester__More_details_in_comments_gvuqt5.mp4'`
I executed the ffmpeg command and got an error. 
I then took out the single quotes around the entire command, and it was successful:
`ffmpeg -i /path/to/gvuqt5_audio -i /path/to/gvuqt5_video -c:v copy -c:a aac -strict experimental /path/to/ReformTheUS_Downtown_NYC_-_8_20pm_-_Police_advancing_and_beating_peaceful_bicyclist_protester__More_details_in_comments_gvuqt5.mp4`
I guess the title didn't have any quotes or special characters. Also, the single quotes around the command had to be taken out.
Hope this helps!

**Comment by aliparlakci** [comment]
I found it. Replace `cmd` with `cmd.split()` the line 54 in src/vreddit.py. It should look like this:
```python
subprocess.call(cmd.split(), stdout=FNULL, stderr=subprocess.STDOUT)
```
I will release a new version with the fix asap.

**Comment by ymgenesis** [comment]
> I found it. Replace `cmd` with `cmd.split()` the line 54 in src/vreddit.py. It should look like this:
> 
> ```python
> subprocess.call(cmd.split(), stdout=FNULL, stderr=subprocess.STDOUT)
> ```
> 
> I will release a new version with the fix asap.

I was JUST looking at that line thinking of how it could be changed, but I'm not a Python person. I just tried your solution and it worked! There's no newline in between ffmpeg posts, but that's just a visual issue if you care about that.

Thanks!

**Comment by aliparlakci** [comment]
Issue is fixed on the version v1.9.1
